presto-main/src/main/java/io/prestosql/operator/scalar/timestamptz/TimestampWithTimeZoneToTimeWithTimezoneCast.java

```diff
@@ -13,44 +13,113 @@
  */
 package io.prestosql.operator.scalar.timestamptz;
 
+import io.prestosql.spi.function.LiteralParameter;
 import io.prestosql.spi.function.LiteralParameters;
 import io.prestosql.spi.function.ScalarOperator;
 import io.prestosql.spi.function.SqlType;
+import io.prestosql.spi.type.LongTimeWithTimeZone;
 import io.prestosql.spi.type.LongTimestampWithTimeZone;
-import io.prestosql.spi.type.StandardTypes;
 import io.prestosql.spi.type.TimeZoneKey;
+import io.prestosql.type.DateTimes;
+
+import java.time.Instant;
 
 import static io.prestosql.spi.function.OperatorType.CAST;
-import static io.prestosql.spi.type.DateTimeEncoding.packDateTimeWithZone;
+import static io.prestosql.spi.type.DateTimeEncoding.packTimeWithTimeZone;
 import static io.prestosql.spi.type.DateTimeEncoding.unpackMillisUtc;
 import static io.prestosql.spi.type.DateTimeEncoding.unpackZoneKey;
 import static io.prestosql.spi.type.TimeZoneKey.getTimeZoneKey;
-import static io.prestosql.type.DateTimes.roundToEpochMillis;
+import static io.prestosql.type.DateTimes.MILLISECONDS_PER_DAY;
+import static io.prestosql.type.DateTimes.PICOSECONDS_PER_DAY;
+import static io.prestosql.type.DateTimes.PICOSECONDS_PER_MILLISECOND;
+import static io.prestosql.type.DateTimes.rescale;
+import static io.prestosql.type.DateTimes.round;
 import static io.prestosql.util.DateTimeZoneIndex.getChronology;
+import static java.lang.Math.floorMod;
 
 @ScalarOperator(CAST)
 public final class TimestampWithTimeZoneToTimeWithTimezoneCast
 {
     private TimestampWithTimeZoneToTimeWithTimezoneCast() {}
 
-    @LiteralParameters("p")
-    @SqlType(StandardTypes.TIME_WITH_TIME_ZONE)
-    public static long cast(@SqlType("timestamp(p) with time zone") long packedEpochMillis)
+    @LiteralParameters({"sourcePrecision", "targetPrecision"})
+    @SqlType("time(targetPrecision) with time zone")
+    public static long shortToShort(
+            @LiteralParameter("targetPrecision") long targetPrecision,
+            @SqlType("timestamp(sourcePrecision) with time zone") long packedTimestamp)
+    {
+        // source precision is <= 3
+        // target precision is <= 9
+        TimeZoneKey zoneKey = unpackZoneKey(packedTimestamp);
+        long epochMillis = getChronology(zoneKey)
+                .getZone()
+                .convertUTCToLocal(unpackMillisUtc(packedTimestamp));
+
+        if (targetPrecision <= 3) {
+            epochMillis = round(epochMillis, (int) (3 - targetPrecision));
+        }
+
+        long nanos = rescale(floorMod(epochMillis, MILLISECONDS_PER_DAY), 3, 9);
+
+        return packTimeWithTimeZone(nanos, DateTimes.getOffsetMinutes(Instant.ofEpochMilli(epochMillis), zoneKey));
```

Comment:
Should this use the zone object that we have above and thus avoid allocation of the `Instant` (and further allocations in the `getOffsetMinutes` method)?

perhaps this should get same offsetMinutes value:
```
getChronology(zoneKey).getZone().getOffset(epochMillis) / MILLISECONDS_PER_MINUTE
```